### ISSUE TITLE
Support for Keycloak v12

### ISSUE THREAD

**Issue by netmikey** [thomasdarimont/embedded-spring-boot-keycloak-server#35]
Keycloak 12.0.0 was released today.
I'm already looking forward for the embedded-spring-boot-keycloak-server project to support it.

**Comment by Michal-Kucera** [comment]
Any update on this? Will the embedded Keycloak get the 12.x support any soon?

**Comment by ccaspanello** [comment]
I too am looking forward to this as well.  Unfortunately upgrading the pom to 12.0.4 is not enough.
<keycloak.version>12.0.4</keycloak.version>

```
2021-03-05 13:26:12.247  INFO 28468 --- [           main] org.keycloak.services                    : KC-SERVICES0050: Initializing master realm
2021-03-05 13:26:12.644  WARN 28468 --- [           main] t.k.e.s.DynamicJndiContextFactoryBuilder : JNDI name not found: java:jboss/ee/concurrency/executor/default
2021-03-05 13:26:12.644  WARN 28468 --- [           main] t.k.e.s.DynamicJndiContextFactoryBuilder : JNDI name not found: java:jboss/ee/concurrency/executor/storage-provider-threads
2021-03-05 13:26:12.648 ERROR 28468 --- [           main] org.keycloak.services                    : KC-SERVICES0011: Failed to add user 'admin' to realm 'master'

java.lang.NullPointerException: null
	at java.base/java.util.concurrent.ConcurrentHashMap.putVal(ConcurrentHashMap.java:1011) ~[na:na]
	at java.base/java.util.concurrent.ConcurrentHashMap.put(ConcurrentHashMap.java:1006) ~[na:na]
	at org.keycloak.executors.DefaultExecutorsProviderFactory.getExecutor(DefaultExecutorsProviderFactory.java:113) ~[keycloak-services-12.0.4.jar:12.0.4]
	at org.keycloak.executors.DefaultExecutorsProviderFactory$1.getExecutor(DefaultExecutorsProviderFactory.java:68) ~[keycloak-services-12.0.4.jar:12.0.4]
	at org.keycloak.utils.ServicesUtils.timeBoundOne(ServicesUtils.java:71) ~[keycloak-services-12.0.4.jar:12.0.4]
	at org.keycloak.storage.AbstractStorageManager.mapEnabledStorageProvidersWithTimeout(AbstractStorageManager.java:135) ~[keycloak-services-12.0.4.jar:12.0.4]
	at org.keycloak.storage.UserStorageManager.getUsersCount(UserStorageManager.java:347) ~[keycloak-services-12.0.4.jar:12.0.4]
	at org.keycloak.models.cache.infinispan.UserCacheSession.getUsersCount(UserCacheSession.java:545) ~[keycloak-model-infinispan-12.0.4.jar:12.0.4]
	at org.keycloak.models.cache.infinispan.UserCacheSession.getUsersCount(UserCacheSession.java:550) ~[keycloak-model-infinispan-12.0.4.jar:12.0.4]
	at org.keycloak.services.managers.ApplianceBootstrap.createMasterRealmUser(ApplianceBootstrap.java:97) ~[keycloak-services-12.0.4.jar:12.0.4]
	at com.github.thomasdarimont.keycloak.embedded.EmbeddedKeycloakApplication.tryCreateMasterRealmAdminUser(EmbeddedKeycloakApplication.java:73) ~[classes/:na]
	at com.github.thomasdarimont.keycloak.embedded.EmbeddedKeycloakApplication.migrateAndBootstrap(EmbeddedKeycloakApplication.java:38) ~[classes/:na]
	at org.keycloak.services.resources.KeycloakApplication$1.run(KeycloakApplication.java:138) ~[keycloak-services-12.0.4.jar:12.0.4]
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:228) ~[keycloak-server-spi-private-12.0.4.jar:12.0.4]
	at org.keycloak.services.resources.KeycloakApplication.startup(KeycloakApplication.java:129) ~[keycloak-services-12.0.4.jar:12.0.4]
	at java.base/java.util.Optional.ifPresent(Optional.java:183) ~[na:na]
	at com.github.thomasdarimont.keycloak.embedded.support.SpringBootPlatform.runStartupHook(SpringBootPlatform.java:36) ~[classes/:na]
	at com.github.thomasdarimont.keycloak.embedded.support.SpringBootPlatform.onApplicationEvent(SpringBootPlatform.java:25) ~[classes/:na]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172) ~[spring-context-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165) ~[spring-context-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139) ~[spring-context-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:404) ~[spring-context-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:361) ~[spring-context-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	at org.springframework.boot.context.event.EventPublishingRunListener.running(EventPublishingRunListener.java:108) ~[spring-boot-2.3.3.RELEASE.jar:2.3.3.RELEASE]
	at org.springframework.boot.SpringApplicationRunListeners.running(SpringApplicationRunListeners.java:77) ~[spring-boot-2.3.3.RELEASE.jar:2.3.3.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:330) ~[spring-boot-2.3.3.RELEASE.jar:2.3.3.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1237) ~[spring-boot-2.3.3.RELEASE.jar:2.3.3.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226) ~[spring-boot-2.3.3.RELEASE.jar:2.3.3.RELEASE]
	at com.github.thomasdarimont.keycloak.embedded.runner.Main.main(Main.java:11) ~[classes/:na]
```

**Comment by jmax01** [comment]
I have this fixed.  I will submit a PR.

**Comment by borag** [comment]
Any news about this ?

**Comment by jmax01** [comment]
> Any news about this ?

Still working on it.

There are two main issues that I fixed:

- `DynamicJndiContextFactoryBuilder` needs to throw when a name isn't found.
Below is my version.  It is a bit more sophisticated then the current implementation as the JNDI objects we expose are configurable via the `KeycloakCustomProperties` :
```java
   public static class KeyCloakInitialContext extends InitialContext {

        public KeyCloakInitialContext(final Hashtable<?, ?> environment) throws NamingException {
            super(environment);
        }

        @Override
        public Object lookup(final Name name) throws NamingException {
            return lookup(name.toString());
        }

        @Override
        public Object lookup(final String name) throws NamingException {
          
           //This is the key part:
            return ofNullable(this.myProps.get(name))
                    .orElseThrow(() -> new NamingException("Name, " + name + ", not found"));

        }

        @Override
        public NameParser getNameParser(final String name) {
            return CompositeName::new;
        }

    }
```
- Added missing providers to config.
12.0.x requires additional providers:

```yaml
        mapStorage:
           provider: ${keycloak.mapStorage.provider:concurrenthashmap}
           #These values are for tests only
           dir: ${project.build.directory:target/keycloak}

        client:
          provider: ${keycloak.client.provider:jpa}
          
        role:
          provider: ${keycloak.role.provider:jpa}
```

**Comment by thomasdarimont** [comment]
Fixed via 82788c2
